File: plugins/sdd-harness/hooks/test_gate.mjs
#!/usr/bin/env node
// Stop gate (project- and stack-agnostic). Before a turn ends, if source files changed
// (working tree / staged / unpushed), run the project's lint+tests and BLOCK on red so the
// turn never finishes with broken code. Auto-detects the stack:
//   - Python (pyproject.toml): `ruff check .` (only if ruff configured) + `pytest -q`
//   - Node   (package.json w/ a "test" script): `<pnpm|yarn|npm> test`
// No matching stack / no source change -> instant no-op. Fail-open on infra errors.
// Respects stop_hook_active (anti-loop) and SDD_HOOKS_DISABLED.
import fs from 'node:fs'
import path from 'node:path'
import { disabled, readStdin, spawnTool } from './_sdd_common.mjs'

const PY = ['.py']
const WEB = ['.ts', '.tsx', '.js', '.jsx', '.vue']

function git(args, cwd) {
  const r = spawnTool('git', args, { cwd, timeout: 20000 })
  return !r.error && r.status === 0 ? (r.stdout || '') : ''
}
function changed(cwd) {
  const s = new Set()
  for (const a of [['diff', '--name-only'], ['diff', '--cached', '--name-only'], ['diff', '--name-only', '@{u}..HEAD']]) {
    for (const l of git(a, cwd).split('\n')) if (l.trim()) s.add(l.trim())
  }
  return [...s]
}
function has(cwd, name) { try { return fs.statSync(path.join(cwd, name)).isFile() } catch { return false } }
function ruffConfigured(cwd) {
  if (has(cwd, 'ruff.toml') || has(cwd, '.ruff.toml')) return true
  try { return fs.readFileSync(path.join(cwd, 'pyproject.toml'), 'utf8').includes('[tool.ruff') } catch { return false }
}
function pm(cwd) { return has(cwd, 'pnpm-lock.yaml') ? 'pnpm' : has(cwd, 'yarn.lock') ? 'yarn' : 'npm' }
function nodeHasTest(cwd) {
  try { return !!((JSON.parse(fs.readFileSync(path.join(cwd, 'package.json'), 'utf8')).scripts || {}).test) } catch { return false }
}
function fail(label, out) { process.stderr.write(`Stop gate FAILED — ${label}:\n` + (out || '').slice(-4000)); return 2 }

function main() {
  if (disabled()) return 0
  let data
  try { data = JSON.parse(readStdin()) } catch { return 0 }
  if (data.stop_hook_active) return 0
  const cwd = data.cwd || process.cwd()
  const ch = changed(cwd)
  const py = ch.some((f) => PY.some((e) => f.endsWith(e)))
  const web = ch.some((f) => WEB.some((e) => f.endsWith(e)))

  if (py && has(cwd, 'pyproject.toml')) {
    if (ruffConfigured(cwd)) {
      const r = spawnTool('python', ['-m', 'ruff', 'check', '.'], { cwd, timeout: 120000 })
      if (!r.error && r.status !== 0) return fail('ruff', (r.stdout || '') + (r.stderr || ''))
    }
    const t = spawnTool('python', ['-m', 'pytest', '-q'], { cwd, timeout: 900000 })
    if (!t.error && t.status !== 0) return fail('pytest', t.stdout || '')
  }

  if (web && nodeHasTest(cwd)) {
    const p = pm(cwd)
    const t = spawnTool(p, ['test'], { cwd, timeout: 900000 })
    if (!t.error && t.status !== 0) return fail(`${p} test`, (t.stdout || '') + (t.stderr || ''))
  }

  return 0
}

process.exit(main())

File: plugins/sdd-harness/hooks/test_gate.py
#!/usr/bin/env python3
"""Stop gate (global, project- and stack-agnostic).

Before a turn ends, if source files changed (working tree / staged / unpushed),
run the project's lint+tests and BLOCK on red so the turn never finishes with
broken code. Auto-detects the stack:
  - Python  (pyproject.toml): `ruff check .` (only if ruff config present) + `pytest -q`
  - Node    (package.json w/ a "test" script): `<pnpm|yarn|npm> test`
No matching stack / no source change -> instant no-op. Fail-open on infra errors.
Respects stop_hook_active (anti-loop) and SDD_HOOKS_DISABLED.
"""
import json
import os
import subprocess
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from _sdd_common import disabled  # noqa: E402

_PY = (".py",)
_WEB = (".ts", ".tsx", ".js", ".jsx", ".vue")


def _git(args, cwd):
    try:
        r = subprocess.run(["git", *args], cwd=cwd, capture_output=True, text=True, timeout=20)
        return r.stdout if r.returncode == 0 else ""
    except Exception:
        return ""


def _changed(cwd):
    out = set()
    out |= set(_git(["diff", "--name-only"], cwd).splitlines())
    out |= set(_git(["diff", "--cached", "--name-only"], cwd).splitlines())
    out |= set(_git(["diff", "--name-only", "@{u}..HEAD"], cwd).splitlines())
    return out


def _run(cmd, cwd, timeout):
    return subprocess.run(cmd, cwd=cwd, capture_output=True, text=True, timeout=timeout)


def _has(cwd, name):
    return os.path.isfile(os.path.join(cwd, name))


def _ruff_configured(cwd):
    if _has(cwd, "ruff.toml") or _has(cwd, ".ruff.toml"):
        return True
    try:
        with open(os.path.join(cwd, "pyproject.toml"), encoding="utf-8") as fh:
            return "[tool.ruff" in fh.read()
    except OSError:
        return False


def _pm(cwd):
    if _has(cwd, "pnpm-lock.yaml"):
        return "pnpm"
    if _has(cwd, "yarn.lock"):
        return "yarn"
    return "npm"


def _node_has_test_script(cwd):
    try:
        with open(os.path.join(cwd, "package.json"), encoding="utf-8") as fh:
            return bool((json.load(fh).get("scripts") or {}).get("test"))
    except Exception:
        return False


def _fail(label, out):
    sys.stderr.write(f"Stop gate FAILED — {label}:\n" + (out or "")[-4000:])
    return 2


def main() -> int:
    if disabled():
        return 0
    try:
        data = json.load(sys.stdin)
    except Exception:
        return 0
    if data.get("stop_hook_active"):
        return 0
    cwd = data.get("cwd") or os.getcwd()
    changed = _changed(cwd)
    py = any(f.endswith(_PY) for f in changed)
    web = any(f.endswith(_WEB) for f in changed)

    if py and _has(cwd, "pyproject.toml"):
        if _ruff_configured(cwd):
            try:
                r = _run(["python", "-m", "ruff", "check", "."], cwd, 120)
            except Exception:
                r = None
            if r is not None and r.returncode != 0:
                return _fail("ruff", (r.stdout or "") + (r.stderr or ""))
        try:
            t = _run(["python", "-m", "pytest", "-q"], cwd, 900)
        except Exception:
            t = None
        if t is not None and t.returncode != 0:
            return _fail("pytest", t.stdout)

    if web and _node_has_test_script(cwd):
        pm = _pm(cwd)
        try:
            t = _run([pm, "test"], cwd, 900)
        except Exception:
            t = None
        if t is not None and t.returncode != 0:
            return _fail(f"{pm} test", (t.stdout or "") + (t.stderr or ""))

    return 0


if __name__ == "__main__":
    sys.exit(main())
